### DISCUSSION TITLE
MAUI CommunityToolkit differences

### DISCUSSION THREAD

**Discussion by symbiogenesis** [TheCodeTraveler/AsyncAwaitBestPractices#146] [Q&A] [answered]
This library provides ValueTask support for commands via AsyncValueCommand, but the MAUI CommunityToolkit doesn't currently offer that.

Is this the main difference? That is kind of a killer feature, but CommunityToolkit works across more products like Blazor, WPF, etc. Makes it hard to choose.

The NotifyCanExecuteChanged events seem to take ages with the CommunityToolkit, and I'm starting to regret porting away.

Since CommunityToolkit is the main competitor, it seems worthwhile to do a writeup about the advantages/disadvantages so that developers can be aware, and not make my mistake.

**Comment by TheCodeTraveler** [comment, marked as answer]
I'm not sure if you are aware of the irony of calling the Community Toolkit a competitor to AsyncAwaitBestPractices. It's not a competitor, rather it's another of my open source projects. 

I'm the lead developer on the .NET MAUI Community Toolkit and I contributed AsyncAwaitBestPractices's AsyncValueCommand implementation to the Xamarin Community Toolkit. 

To answer your question, the path forward for .NET MAUI developers is to use CommunityToolkit.MVVM's RelayCommand: https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/relaycommand

It's the most performant implementation of an Async Command and its implementation using Source Generators will save you from writing much boiler plate code. 

Highly recommend it 💯

**Comment by andrekiba** [comment]
Hi @brminnick and thanks in advance.
May I ask why the implementation of AsyncRelayCommand (and variations) in CommunityToolkit is a `sealed` class??
And why in your AsyncAwaitBestPractices the default for AsyncCommand is `ContinueOnCapturedContext = false` when, I suppose, in the main scenario where could be used the context is most of the time very important??
I'm asking because I have a custom implementation that will re-evaluate the `canExecute` based on the fact that some dependency properties may changes (maybe using `Fody.PropertyChanged` also).
As a base class for my custom implementation I'm using yours AsyncAwaitBestPractices.AsyncCommand

**Reply by TheCodeTraveler** [reply]
Hi @andrekiba! The default value for `ContinuedOnCapturedContext` is `false` because that is the most performant setting for the most common usage scenario of `AsyncCommand`: Most developers will use it within an MVVM architecture that fires `ICommand.Execute(object)` triggered by a user input (eg clicking a button). 

I've made it `sealed` also for performance reasons. If you'd like to extend the functionality of the library, we can add new features and publish them to NuGet. You're also welcome to copy/paste the source code and use it as you'd like.

**Reply by andrekiba** [reply]
Ciao @brminnick thank you for your response.
I have to say that I’m not totally agree 😅
Ok for the performances…
But in a framework (Maui, Xamarin, Forms…) where the synchronization context matters why reverse the default awaitable behavior?
Ok that there is the marshaling of the `PropertyChanged` on the UI thread (perhaps maybe not everywhere). Allow to switch the thread…to then marshal on the ui one?
There is so much confusion on this topic (in general around async-await).
The mvvn toolkit `AsyncRelayCommand` does the same? Thanks!
